lms/djangoapps/courseware/access.py

```diff
@@ -562,6 +565,20 @@ def can_load():
         if not group_access_response:
             return group_access_response
 
+        if is_integrity_signature_enabled():
+            # If the user's enrollment mode is non-audit, check if they have an integrity
+            # signature for the course, and deny access if not
+            course_id = str(course_key)
+            enrollment = get_enrollment(user.username, course_id)
+            if enrollment and enrollment.get('mode') != 'audit':
```

Comment:
It's better to ensure the modes should be part of [VERIFIED_MODES](https://github.com/edx/edx-platform/blob/master/common/djangoapps/course_modes/models.py#L169). See function https://github.com/edx/edx-platform/blob/master/common/djangoapps/course_modes/models.py#L621